incubator/hnc/internal/validators/object_test.go

```diff
@@ -5,14 +5,65 @@ import (
 	"testing"
 
 	. "github.com/onsi/gomega"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
 	api "sigs.k8s.io/multi-tenancy/incubator/hnc/api/v1alpha1"
 	"sigs.k8s.io/multi-tenancy/incubator/hnc/internal/forest"
 	"sigs.k8s.io/multi-tenancy/incubator/hnc/internal/metadata"
+	"sigs.k8s.io/multi-tenancy/incubator/hnc/internal/reconcilers"
 )
 
+func TestType(t *testing.T) {
+	or := &reconcilers.ObjectReconciler{
+		GVK:  schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Secret"},
+		Mode: api.Propagate,
+	}
+	f := forest.NewForest()
+	f.AddTypeSyncer(or)
+	l := zap.Logger(false)
+	o := &Object{Forest: f, Log: l}
+
+	tests := []struct {
+		name    string
+		version string
+		kind    string
+		fail    bool
+	}{{
+		name:    "Deny GroupVersionKind in the propagate mode",
+		version: "v1",
+		kind:    "Secret",
+		fail:    true,
```

Comment:
It's unclear to me _why_ the admission controller is returning a "denied" here (maybe `allow` or `deny` would be a better term here). Is it because we're not actually including a `Secret` object, so the call to `Decode` will fail? So basically, we're passing in a request that's _guaranteed_ to fail _unless_ we skip it because the type isn't registered?

If so, I think that deserves a lot of commenting somehwere, maybe on line 53 or so.

Reply:
Added comments in validators/object.go#41 and object_test.go#21. 

TestType only tests the types from the admission requests. There is no object in the admission requests for all the test cases. The requests are supposed to fail with "BadRequest" since the object cannot be decoded. However, with early exit for non-propagate-mode types, they should return "allow" before the object gets decoded.